File: frontend/src/types/post.ts
import type { Classification } from './common'

export type PostCategory =
  | 'huan_luyen'
  | 'dan_van'
  | 'khen_thuong'
  | 'guong_nguoi_tot'
  | 'hoat_dong_don_vi'
  | 'cong_tac_dang'
  | 'thong_tin_lien_lac'
  | 'su_kien_le_ky_niem'
export type PostStatus = 'nhap' | 'cho_duyet' | 'da_duyet' | 'tra_lai'

export const POST_CATEGORY_LABELS: Record<PostCategory, string> = {
  huan_luyen: 'Huấn luyện',
  dan_van: 'Dân vận – Đối ngoại biên phòng',
  khen_thuong: 'Thi đua – Khen thưởng',
  guong_nguoi_tot: 'Gương người tốt việc tốt',
  hoat_dong_don_vi: 'Hoạt động đơn vị',
  cong_tac_dang: 'Công tác Đảng – công tác chính trị',
  thong_tin_lien_lac: 'Thông tin liên lạc – chuyên môn',
  su_kien_le_ky_niem: 'Sự kiện – Lễ, kỷ niệm',
}

export const POST_STATUS_LABELS: Record<PostStatus, string> = {
  nhap: 'Bản nháp',
  cho_duyet: 'Chờ duyệt',
  da_duyet: 'Đã duyệt',
  tra_lai: 'Trả lại',
}

export interface Post {
  id: number
  title: string
  summary: string | null
  slug: string | null
  category: PostCategory
  content: string
  tags: string[]
  cover_image_url: string | null
  classification: Classification
  is_featured: boolean
  status: PostStatus
  review_note: string | null
  reviewed_by_id: number | null
  reviewed_at: string | null
  author_id: number
  author_full_name: string
  created_at: string
}

export interface PostCreate {
  title: string
  summary?: string | null
  /** Để trống → backend tự sinh từ tiêu đề (bỏ dấu tiếng Việt), trùng thì thêm hậu tố. */
  slug?: string | null
  category: PostCategory
  content: string
  tags: string[]
  cover_image_url?: string | null
  classification: Classification
  is_featured: boolean
}

export interface PostReview {
  status: 'da_duyet' | 'tra_lai'
  review_note?: string | null
}

File: backend/app/schemas/post.py
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.common import Classification

PostCategory = Literal[
    "huan_luyen",
    "dan_van",
    "khen_thuong",
    "guong_nguoi_tot",
    "hoat_dong_don_vi",
    "cong_tac_dang",
    "thong_tin_lien_lac",
    "su_kien_le_ky_niem",
]
# nhap = ban nhap (chua gui duyet); cho_duyet -> da_duyet / tra_lai
PostStatus = Literal["nhap", "cho_duyet", "da_duyet", "tra_lai"]

MAX_TAGS = 20
MAX_TAG_LEN = 40


class PostCreate(BaseModel):
    title: str = Field(..., max_length=255)
    summary: Optional[str] = Field(None, max_length=500)
    # De trong -> backend tu sinh tu title (bo dau tieng Viet). Bi trung -> them hau to.
    slug: Optional[str] = Field(None, max_length=255)
    category: PostCategory
    content: str
    tags: list[str] = Field(default_factory=list)
    cover_image_url: Optional[str] = Field(None, max_length=500)
    classification: Classification = "noi_bo"
    is_featured: bool = False

    @field_validator("tags")
    @classmethod
    def _clean_tags(cls, raw: list[str]) -> list[str]:
        seen: list[str] = []
        for tag in raw:
            t = (tag or "").strip()[:MAX_TAG_LEN]
            if t and t.lower() not in {s.lower() for s in seen}:
                seen.append(t)
        return seen[:MAX_TAGS]


class PostReview(BaseModel):
    # da_duyet = duyet dang; tra_lai = tra ve cho tac gia sua
    status: Literal["da_duyet", "tra_lai"]
    review_note: Optional[str] = Field(None, max_length=500)


class PostOut(BaseModel):
    id: int
    title: str
    summary: Optional[str]
    slug: Optional[str]
    category: PostCategory
    content: str
    tags: list[str]
    cover_image_url: Optional[str]
    classification: Classification
    is_featured: bool
    status: PostStatus
    review_note: Optional[str]
    reviewed_by_id: Optional[int]
    reviewed_at: Optional[datetime]
    author_id: int
    author_full_name: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)
